keps/sig-apps/4443-configurable-pod-failure-policy-reasons/README.md

```diff
@@ -0,0 +1,957 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes to extend the Job API by adding an optional `SetConditionReason` field to `PodFailurePolicyRule`, which if specified, would be included as the reason in the `JobFailed` condition upon Job failure triggered by a `PodFailurePolicy`.
+
+## Motivation
+
+Higher level APIs, such as [JobSet](https://sigs.k8s.io/jobset), use the Job API as a building block for orchestrating large, distributed workloads.
+These higher level APIs need to be able to distinguish between different types of Job failures in order to make informed decisions about how to react
+to them. Currently, no mechanism exists in the Job API to propagate granular failure reason information (e.g., container exit codes) up to be 
+programmatically consumed by higher level software managing Jobs. A `PodFailurePolicy` can be configured to add a condition reason of `PodFailurePolicy`
+to the `JobFailed` condition added to the Job when it fails, but different pod failure policies targeting different container exit codes all use the
+same condition reason of `PodFailurePolicy`. This prevents higher level APIs like JobSet from distinguishing them and being able to take different
+actions depending on the type of Job failure that occurred.
+
+For a concrete use case, see the JobSet [Configurable Failury Policy KEP](https://github.com/kubernetes-sigs/jobset/pull/381) which illuminated the need for more granular pod failure policy reasons.
+
+### Goals
+
+For pod failure policies to be able communicate different failure types to higher level APIs.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+The proposal is to add an optional `SetConditionReason` field to the `PodFailurePolicyRule` that is used
+to set the `Reason` field on the `JobFailed` condition when a Job fails because of a pod failure policy rule match.
+
+If unset, it will default to `PodFailurePolicy`, which is the current [reason](https://sourcegraph.com/github.com/kubernetes/kubernetes@6a4e93e776a35d14a61244185c848c3b5832621c/-/blob/staging/src/k8s.io/api/batch/v1/types.go?L542) the Job
```

Comment:
nit: I would probably just link to the code. Not that familar with sourcegraph but it would be nice to have these links work for docs purposes.

Reply:
Not sure if we can link to specific commits but `git.k8s.io/kubernetes/staging/src/k8s.io/api/batch/v1/types.go#L542
` might be preferable.